### ISSUE TITLE
 Replacing CE by KL in the inner maximization

### ISSUE THREAD

**Issue by mangoerya** [YisenWang/MART#5]
Hi Yisen,
Thank you for your sharing!
The paper [https://arxiv.org/pdf/2002.11242.pdf](url) evaluates MART and the reported robust accuracy under C&W drops by ~6% compared to the accuracy under PGD attack, while in your paper the reported accuracy under those two attacks are about the same. I run your code and there's a gap between accuracy under PGD and under other attacks when evaluated using [https://github.com/fra31/auto-attack](url). 
Besides, when I replace CE by KL in the inner maximization, the accuracy under PGD drops by ~3% and the gap between PGD and other attacks is down to ~3%.  However, the accuracy reported in your paper does not vary much when replacing CE by KL in the inner maximization. Do you have any thoughts on what may go wrong?

**Comment by YisenWang** [comment]
Hi, thanks for your interest for our work. 

For your first question, the attack parameters for CW in these two papers are different. We used different kappa parameters. 

For your second question, the accuracy under PGD and auto-attack should be different, because auto-attack ensembles four attacks. 

For your third question, would you please indicate your experiment settings before I can locate your questions?

**Comment by mangoerya** [comment]
Thanks for the quick reply!
I use the default settings in `train_resnet.py` and change only the inner max loss function in `mart.py` (line 25) from CE to KL. I then evaluate the two models using auto-attack, and below are accuracies I get:
- CE for inner max:
        robust accuracy after APGD-CE: 52.06% / after APGD-T: 47.18% (trained for 120 epochs)
        robust accuracy after APGD-CE: 53.49% / after APGD-T: 47.90% (early stop)
- KL for inner max:
        robust accuracy after APGD-CE: 50.79% / after APGD-T: 47.55% (trained for 120 epochs)
        robust accuracy after APGD-CE: 51.75% / after APGD-T: 48.29% (early stop)
The robust accuracy under PGD is increased when replacing KL by CE in the inner max, and the gap between the accuracy under PGD and under ensemble attack become wider.

**Comment by YisenWang** [comment]
Actually, we did not use auto-attack in our paper because it is published after ours. But I think your experimental phenomenon is interesting, which can be investigated further. For example, the correctly classified example by PGD may be misclassified by APGD, then can design an adaptive attack during adversarial training.
